### PR TITLE
Global helper methods

### PR DESCRIPTION
**Pull request by thekanbo**

### CODE REVIEW

File: helpers.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+const format  = require('format');
+const _       = require('lodash');
+
+// Using getMessages will automatically use already cached messages when available,
+//     and will fetch more messages only if needed. This way we don't have to rely
+//     on the promise function structure every time we want to retrieve a message
+exports.getMessages = async function getMessages(message) {
+
+  // Current messages in the Discord client's cache
+  let cachedMessages = message.channel.messages.array()[0].channel.messages;
```
Comment (skeltonath) [inline helpers.js]: what does is the structure of that object?

weird that you have to do `.array()[0].channel.messages`
will it always be at the first spot in the array? and will there always be a `[0]`?

File: helpers.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  
+  // If there are only a couple of items in the cache...
+  if (cachedMessages.array().length < 10) {
```
Comment (skeltonath) [inline helpers.js]: do messages fall out of that cache?

File: helpers.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
+const log4js  = require('log4js');
+const LOG     = log4js.getLogger('helpers');
+const format  = require('format');
+const _       = require('lodash');
+
+// Using getMessages will automatically use already cached messages when available,
+//     and will fetch more messages only if needed. This way we don't have to rely
+//     on the promise function structure every time we want to retrieve a message
+exports.getMessages = async function getMessages(message) {
```
Comment (skeltonath) [inline helpers.js]: if we start using this in handlers, how long will it take a new message to enter the cache?

is this just a way for handlers that need messages for sampling to get them quickly without a network request? i guess those kinds don't need the _latest_ messages ASAP.

File: helpers.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    cachedMessages = await message.channel.fetchMessages({limit: 100});
+  }
+
+  return cachedMessages;
+}
+
+
+// Randomly look through all results until we get a message that 
+//     isnt bot generated or a bot command
+exports.filterMessages = function filterMessage(messages) {
```
Comment (skeltonath) [inline helpers.js]: why don't we call this like `getUserMessage` or `getRandomNonBotMessage` haha? `filterMessages is a little ambiguous.

File: helpers.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+exports.getEmoji = function getEmoji(emoji) {
+  return new Promise((resolve, reject) => {
+
+  });
+}
+
+// Simple array shuffle
+function shuffle(o) {
```
Comment (skeltonath) [inline helpers.js]: lodash has an array shuffle method

`_.shuffle()`

File: helpers.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  let i = 1;
+  while ( (found.author.bot || found.content.startsWith('.')) && i < newMessages.length) {
+    found = newMessages[i];
+    i++;
+  }
+
+  return found;
+}
+
+exports.getEmoji = function getEmoji(emoji) {
```
Comment (skeltonath) [inline helpers.js]: nothing here?

File: helpers.js (hunk 3 above)
Comment (thekanbo) [reply helpers.js]: When the client starts, there are no messages in the cache, which is why when we looked earlier, it seemed like the 'messages' data object only had a single message. After the client starts, channel messages will fill that cache automatically, and running `fetchMessages()` backfills it. Using this method, we will only need to run `fetchMessages` once per instance to backfill the memory, and all future messages are automatically loaded into the cache memory.

File: helpers.js (hunk 2 above)
Comment (thekanbo) [reply helpers.js]: Not sure if/when messages fall out; this check in particular is arbitrary and just to backfill if it hasn't seen many messages yet

File: helpers.js (hunk 6 above)
Comment (thekanbo) [reply helpers.js]: https://github.com/skeltonath/schreef/issues/43

**Comment by thekanbo** [comment]
Leaving out formatting issues until re-merge with master; will lint after resolving conficts

**Pull request merged**